server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -711,7 +712,12 @@ private static void parseCopy(String field, ParseContext context) throws IOExcep
                     return new Tuple<>(pathsAdded, parent);
                 } else {
                     //objects are created under properties even with dynamic: runtime, as the runtime section only holds leaf fields
-                    mapper = (ObjectMapper) dynamic.getDynamicFieldsBuilder().createDynamicObjectMapper(context, paths[i]);
+                    final Mapper fieldMapper = dynamic.getDynamicFieldsBuilder().createDynamicObjectMapper(context, paths[i]);
+                    if (fieldMapper instanceof ObjectMapper == false) {
+                        throw new MapperParsingException("Field [" + context.path().pathAsText(paths[i]) + "] must be an object; " +
+                            "but it's configured as [" + fieldMapper.typeName() + "] in dynamic templates");
```

Comment:
Question: can this happen in reality? If so do we have a test for it? I am wondering if this became relevant with the previous iteration but it may no longer be the case now that hints are separate from types. Maybe we should rather have an assertion in createDynamicObjectMapper and cast there so that its return type can be the right one? This is unrelated to your PR though :)

Reply:
I still wonder about this added conditional ;)

Reply:
I added a test for this in https://github.com/elastic/elasticsearch/pull/69948/commits/a6f99710eda8c035acad03789f0684047509e36f.